### DISCUSSION TITLE
some doubts

### DISCUSSION THREAD

**Discussion by empereira** [nfrastack/container-gcds#4] [Ideas]
hi,

Where does the value for the variable $GCDS_XML_FILE come from?

**Comment by tiredofit** [comment]
That's the default GCDS configuration file, shown here in defaults.
https://github.com/tiredofit/docker-gcds/blob/e8a4e8debf877c3d741f9671a482b7f875fdb050/install/assets/defaults/10-gcds#L9

Essentially it gets copied in and out per container start from /assets/config/(whatever your filename is of) $CONFIG_FILE into that location for operation on Oauth/LDAP/Encryption.

**Comment by empereira** [comment]
Thanks for the reply!!!

When I run the command `sh auth.sh` it returns the following error:

```
[tiredofit/gcds:5.1.6 11:09:31 /gcds] $ sh auth.sh 
auth.sh: 2: source: not found
auth.sh: 3: source: not found
auth.sh: 5: print_notice: not found
***               GOOGLE CLOUD DIR SYNC OAUTH STEP
***
***         This script will connect to Google's Servers and provide you with a link to visit using your local web browser. 
***
***
auth.sh: 11: read: Illegal option -s
no argument for:c
usage: upgrade-config [-exportkeys filename [password]] [-ldapuser
       username] [-p] [-smtppassword password] [-v] [-V] [-importkeys filename
       [password]] [-smtpusername username] [-testldap] [-ldappassword password]
       [-c filename] [-Oauth domain] [-h] [-testgoogleapps]
```

I ran the commands manually, and when running `/gcds/upgrade-config -Oauth $DOMAIN -c $GCDS_XML_FILE` it returns the error below. I am currently using a windows 7 with GCDS working. I copied the xml I use in production.

```
[2022-08-26 11:27:56,675+0000] [main] [FATAL] [usersyncapp.cli.UpgradeConfig] Unable to load the configuration from the specified file "/gcds/gcds/gcds_conf.xml" - smtpAuthPasswordEncrypted contains an invalid encrypted value
```

You must delete the following entries in config.xml:

- smtpAuthPasswordEncrypted
- oAuth2RefreshToken
- encryptedSslProxyPassword

https://support.google.com/a/answer/7177266#nogui&zippy=%2Chow-do-i-authorize-gcds-on-a-machine-without-a-gui

But, the error below still occurred:

```
[main] [ERROR] [config.loaders.LdapConfigLoader] com.google.usersyncapp.util.EncryptionManager$InvalidCipherTextException: Invalid encryption parameters. Salt/Iteration/Initialization Vector
Please open the following address in your browser:
```

**Reply by tiredofit** [reply]
Can you try running that with bash instead of sh? I just looked and the +x bit is not set on that script. I will push a new release that includes it. You will get different results.

**Reply by tiredofit** [reply]
New release tagged as `5.1.7`

**Comment by empereira** [comment]
Ok. Now it runs correctly with bash.

Is this URL correct?

```
$ bash auth.sh 
2022-08-26.14:40:49 [NOTICE] ** [container] 
***               GOOGLE CLOUD DIR SYNC OAUTH STEP
***
***         This script will connect to Google's Servers and provide you with a link to visit using your local web browser. 
***
***
Press any key to continue...
[2022-08-26 14:40:50,905+0000] [main] [WARN] [usersyncapp.cli.UpgradeConfig] Upgrading configuration file "/gcds/gcds_conf.xml" to most recent version (will save a backup).
Please open the following address in your browser:
  https://accounts.google.com/o/oauth2/auth?client_id=xxxxxxxxxxx.apps.googleusercontent.com&redirect_uri=urn:ietf:wg:oauth:2.0:oob&response_type=code&scope=https://www.googleapis.com/auth/admin.directory.resource.calendar%20https://www.googleapis.com/auth/apps.licensing%20https://www.googleapis.com/auth/apps.groups.settings%20https://www.googleapis.com/auth/admin.directory.orgunit%20https://www.googleapis.com/auth/plus.me%20https://www.google.com/m8/feeds/%20https://www.googleapis.com/auth/admin.directory.user%20https://www.googleapis.com/auth/admin.directory.userschema%20https://www.googleapis.com/auth/admin.directory.group
Please enter code:
```
I get the error in the image below when I open it!

![image](https://user-images.githubusercontent.com/29874574/186930748-d6325481-1522-4ffa-b722-2c14068a2cc9.png)

**Reply by tiredofit** [reply]
Yes, URL is correct and is issued from GCDS itself, not baked into the image from my end. I just auth'ed on a system and worked ok.

Try visiting the URL with incognito mode?

**Comment by empereira** [comment]
I tried in incognito mode and the same error occurs as in the image.

Here some info about the error.
https://support.google.com/a/answer/1263028?product_name=UnuFlow&hl=en&visit_id=637971225953720554-470443744&rd=1&src=supportwidget0&hl=en#zippy=

**Comment by empereira** [comment]
Hello Dave,

Any tips on how I can solve these problems? Which version of GCDS are you using?

**Reply by tiredofit** [reply]
Sorry I don't have any more advice as I haven't encountered it. I'm Just running `:latest` as the tool auto upgrades the configuration to the latest version on start. I have been using this regularly since early 2017 and not seen the same issues.

**Comment by empereira** [comment]
Did you configure the xml file on windows before moving to linux? 

Did you configure the xml file differently? As I posted just above, where google docs asks you to remove some entries from this file?

**Reply by tiredofit** [reply]
The image actually removes the correct pieces if you first use a GUI based configuration tool. I don't use Windows but just used the Java GUI which is likely the exact same as what you are using.

I originally started and used the GUI in 2017, and used it once last year just to make sure I was aware of all the latest features and it seemed to  be fine. 

My work strips the encryption password and resets the Oauth token storing it persistently so that the container can survive reboots - Its a fairly brute force way of hacking away at the XML but once working, keeps working. (That would be option 2 in that link)

I do my updates to the config manually now as my config is solid, and any changes would be to add or remove an OU which doesn't require GUIs.

**Comment by empereira** [comment]
Ok.

Does your xml have these two options below?

nonAddressPrimaryKeyMapFile
passwordTimestampFile

**Reply by tiredofit** [reply]
Yes, in the configuration, however I only have one of the TSVs listed inside the `/gcds` folder. GADS_nonAddressPrimaryKeyFile.tsv

**Reply by empereira** [reply]
Ok. I haven't gotten the authorization yet. I always get the error mentioned above.

I will keep looking for an alternative.

Thank you very much for all your time!!!

**Comment by empereira** [comment]
Hello Dave,

I installed the latest version of gcds on my ubuntu 20.x desktop and via GUI the authorization worked.

I tried via ./upgrade-config and the same error occurred.

Is there a problem in one of the URLs of the generated link?

**Reply by tiredofit** [reply]
I am not entirely sure- The link is generated by GCDS itself. I suppose you could try to export and import keys and see how that works for you?

**Reply by empereira** [reply]
I exported and imported the file.

Is that all? I don't need to change anything in the xml file?
